m3/util.go

```diff
@@ -0,0 +1,61 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package m3
+
+import (
+	"bytes"
+)
+
+const (
+	defaultReplacementCharacter = '_'
+)
+
+var (
+	validRangeCharacters  = [][2]int{{int('a'), int('z')}, {int('A'), int('Z')}, {int('0'), int('9')}}
+	validSingleCharacters = []rune{'-', '_'}
+)
+
+func sanitizeMetricIdentifiers(value string) string {
+	buf := bytes.NewBuffer(make([]byte, 0, len(value)))
```

Comment:
To avoid allocations in the general case we can lazily construct this buffering similar to how some parts of the Go stdlib does replacements:
```go
var buf *bytes.Buffer
for i, ch := range value {
  // ... use existing code to calculate validCurr ...

  if validCurr {
    if buf == nil {
      // Still don't need to deviate from string
      continue
    }
    buf.WriteRune(ch)
    continue
  }

  if buf == nil {
    // Definitely have deviated, need to initialize buf and backfill
    buf = bytes.NewBuffer(make([]byte, 0, len(value)))
    if i > 0 {
      buf.WriteString(value[:i-1])
    }
  }
  buf.WriteRune(defaultReplacementCharacter)
}

func lazyInitBuf(buf *bytes.Buffer) 
```